### PR TITLE
AArch64: Implementation of Table Evaluator

### PR DESCRIPTION
**Pull request by marufunb**
This commit implements Table Evaluator for openj9.

Signed-off-by: marufunb <mrahma15@unb.ca>

### CODE REVIEW

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 1]
```diff
@@ -480,7 +480,81 @@ TR::Register *
...
+         generateConditionalBranchInstruction(cg, TR::InstOpCode:: b_cond, node, node->getChild(2+i)->getBranchDestination()->getNode()->getLabel(), TR::CC_EQ);
+         }
+      generateConditionalBranchInstruction(cg, TR::InstOpCode:: b_cond, node, secondChild->getBranchDestination()->getNode()->getLabel(), TR::CC_AL, conditions);
+      }
+   else
+      {
+      TR::RealRegister *machineIP        = cg->machine()->getRealRegister(TR::RealRegister::gr31);
+      uint32_t      base, rotate;
+
+      if (!constantIsImmed8r(numBranchTableEntries, &base, &rotate))
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ControlFlowEvaluator.cpp]: This is the wrong function to call if you're using a CCMP immediate instruction (which only supports a 5-bit immediate).  Please read the description in section 6.2.45 of the ARM architecture manual.

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 2]
```diff
@@ -480,7 +480,81 @@ TR::Register *
...
+         {
+         TR::Register *tempRegister = cg->allocateRegister();
+         loadConstant64(cg, node, numBranchTableEntries, tempRegister);
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ControlFlowEvaluator.cpp]: Why a 64-bit constant, not 32?

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 3]
```diff
@@ -480,7 +480,81 @@ TR::Register *
...
+         generateSrc2Instruction(cg, TR::InstOpCode::ccmpw, node, selectorReg, tempRegister);
+         }
+      else
+         {
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ccmpimmw, node, rotate, selectorReg, base);
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ControlFlowEvaluator.cpp]: This is not the correct form of this instruction.  See 6.2.45.

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 4]
```diff
@@ -480,7 +480,81 @@ TR::Register *
...
+         }
+      generateConditionalBranchInstruction(cg, TR::InstOpCode:: b_cond, node, secondChild->getBranchDestination()->getNode()->getLabel(), TR::CC_CS);
+
+      // code looks like:dd
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ControlFlowEvaluator.cpp]: Comment needs to be modified to match your code.

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 5]
```diff
@@ -480,7 +480,81 @@ TR::Register *
 OMR::ARM64::TreeEvaluator::tableEvaluator(TR::Node *node, TR::CodeGenerator *cg)
 	{
 	// TODO:ARM64: Enable TR::TreeEvaluator::tableEvaluator in compiler/aarch64/codegen/TreeEvaluatorTable.hpp when Implemented.
-	return OMR::ARM64::TreeEvaluator::unImpOpEvaluator(node, cg);
+   int32_t       total = node->getNumChildren();
+   int32_t       numBranchTableEntries = total - 2;
+   TR::Register   *selectorReg           = cg->evaluate(node->getFirstChild());
+   int32_t       i;
+   TR::Node       *secondChild = node->getSecondChild();
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ControlFlowEvaluator.cpp]: Please choose a better name for this child to make the code more readable.  Perhaps the `defaultChild`?

File: compiler/aarch64/codegen/OMRInstruction.cpp [hunk 6]
```diff
@@ -27,7 +27,32 @@
 #include "codegen/RegisterDependency.hpp"
 
 namespace TR { class Register; }
+bool constantIsImmed8r(int32_t intValue, uint32_t *base, uint32_t *rotate)
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/OMRInstruction.cpp]: I don't think this function applies to this PR.

**Comment by marufunb** [comment]
@0xdaryl I have made those changes according to your feedback. Can you please have a look on this new changes? Thanks.

**Comment by aarongraham9** [comment]
@marufunb/@0xdaryl should this be against omr or against openj9?

**Comment by aarongraham9** [comment]
What's the status of this one @marufunb?

**Comment by marufunb** [comment]
No feedback from Daryl.

Thanks,
Md. Mahbubur Rahman (Maruf)
Graduate Academic Assistant at
IBM/UNB Centre of Advanced Studies- Atlantic (CASA)
Phone: 5064714698
Faculty of Computer Science,
University of New Brunswick

________________________________

**Comment by knn-k** [comment]
My temporary code in #4163 has caused a merged conflict.  Please rebase.

**Comment by marufunb** [comment]
This PR (#3858) has been superseded by #4214. Closing.

**Pull request closed without merging**